File: tests/scalers/azure-log-analytics.test.ts
import * as fs from 'fs'
import * as sh from 'shelljs'
import * as tmp from 'tmp'
import test from 'ava'

const defaultNamespace = 'azure-log-analytics-test'
const la_workspace_id = process.env['TEST_LOG_ANALYTICS_WORKSPACE_ID']
const sp_id = process.env['AZURE_SP_ID']
const sp_key = process.env['AZURE_SP_KEY']
const sp_tenant = process.env['AZURE_SP_TENANT']

test.before(t => {
  if (!la_workspace_id || !sp_id || !sp_key || !sp_tenant) {
    t.fail('Connection parameter for LA scaler was not resolved')
  }

  sh.config.silent = true

  sh.exec(`kubectl create namespace ${defaultNamespace}`)
  t.is(
    0,
    sh.exec(`kubectl apply -f ${createYamlFile(deployYaml)}`).code,
    'creating a deployment should work.'
  )
})

test.serial('Deployment should have 0 replicas on start', t => {
  const replicaCount = sh.exec(
    `kubectl get statefulset.apps/test-deployment --namespace ${defaultNamespace} -o jsonpath="{.spec.replicas}"`
  ).stdout
  t.is(replicaCount, '0', 'replica count should start out as 0')
})

test.serial.cb('Deployment should scale to 2 replicas', t => {
  t.is(
    0,
    sh.exec(`kubectl apply -f ${createYamlFile(secretYaml)}`).code,
    'creating a scaled object should work.'
  )
  t.is(
    0,
    sh.exec(`kubectl apply -f ${createYamlFile(triggerAuthYaml)}`).code,
    'creating a scaled object should work.'
  )
  t.is(
    0,
    sh.exec(`kubectl apply -f ${createYamlFile(scaledObjectYaml)}`).code,
    'creating a scaled object should work.'
  )

  //Checking replicas
  let replicaCount = '0'
  for (let i = 0; i < 180 && replicaCount !== '2'; i++) {
    replicaCount = sh.exec(
      `kubectl get statefulset.apps/test-deployment --namespace ${defaultNamespace} -o jsonpath="{.spec.replicas}"`
    ).stdout
    if (replicaCount !== '2') {
      sh.exec('sleep 1s')
    }
  }

  t.is('2', replicaCount, 'Replica count should be 2 after some time')

  t.end()
})

test.serial.cb('Deployment should scale to 0 replicas', t => {
  //Let's change a query to scale down
  t.is(
    0,
    sh.exec(`kubectl patch scaledobject test-scaledobject -n ${defaultNamespace} --type='json' -p='[{"op": "replace", "path": "/spec/triggers/0/metadata/query", "value":"let x = 0; let y = 1; print MetricValue = x, Threshold = y;"}]'`).code,
    'changing a scaled object should work.'
  )

  //Checking replicas
  let replicaCount = '0'
  for (let i = 0; i < 180 && replicaCount !== '0'; i++) {
    replicaCount = sh.exec(
      `kubectl get statefulset.apps/test-deployment --namespace ${defaultNamespace} -o jsonpath="{.spec.replicas}"`
    ).stdout
    if (replicaCount !== '0') {
      sh.exec('sleep 1s')
    }
  }

  t.is('0', replicaCount, 'Replica count should be 0 after some time')

  t.end()
})

test.after.always.cb('clean up deployment', t => {
  const resources = [
    'scaledobject.keda.sh/test-scaledobject',
    'statefulset.apps/test-deployment',
  ]

  for (const resource of resources) {
    sh.exec(`kubectl delete ${resource} --namespace ${defaultNamespace}`)
  }
  sh.exec(`kubectl delete namespace ${defaultNamespace}`)

  t.end()
})

function createYamlFile(yaml: string) {
  const tmpFile = tmp.fileSync()
  fs.writeFileSync(tmpFile.name, yaml
    .replace('{{NAMESPACE}}', defaultNamespace)
    .replace('{{TENANT_ID}}', sp_tenant)
    .replace('{{CLIENT_ID}}', Buffer.from(sp_id).toString('base64'))
    .replace('{{CLIENT_SECRET}}', Buffer.from(sp_key).toString('base64'))
    .replace('{{WORKSPACE_ID}}', la_workspace_id))

  return tmpFile.name
}

const deployYaml = `apiVersion: v1
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: test-deployment
  namespace: {{NAMESPACE}}
spec:
  serviceName: "test-deployment"
  replicas: 0
  selector:
    matchLabels:
      app: test-deployment
  template:
    metadata:
      labels:
        app: test-deployment
    spec:
      containers:
      - name: nginx
        image: nginx:1.16.1
        ports:
        - containerPort: 80
`
const secretYaml = `apiVersion: v1
kind: Secret
metadata:
  name: test-scaledobject-secret
  namespace: {{NAMESPACE}}
type: Opaque
data:
  la-clientId: {{CLIENT_ID}}
  la-clientSecret: {{CLIENT_SECRET}}`

const triggerAuthYaml = `apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: test-scaledobject-trigger-auth
  namespace: {{NAMESPACE}}
spec:
  secretTargetRef:
    - parameter: clientId
      name: test-scaledobject-secret
      key: la-clientId
    - parameter: clientSecret
      name: test-scaledobject-secret
      key: la-clientSecret`

const scaledObjectYaml = `apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: test-scaledobject
  namespace: {{NAMESPACE}}
spec:
  scaleTargetRef:
    kind: StatefulSet
    name: test-deployment
  pollingInterval: 5
  cooldownPeriod: 5
  maxReplicaCount: 2
  triggers:
    - type: azure-log-analytics
      metadata:
        tenantId: "{{TENANT_ID}}"
        workspaceId: "{{WORKSPACE_ID}}"
        query: "let x = 10; let y = 1; print MetricValue = x, Threshold = y;"
        threshold: "1"
      authenticationRef:
        name: test-scaledobject-trigger-auth
`

File: tests/scalers_go/azure_log_analytics/azure_log_analytics_test.go
//go:build e2e
// +build e2e

package azure_log_analytics_test

import (
	"encoding/base64"
	"fmt"
	"os"
	"testing"

	"github.com/joho/godotenv"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"k8s.io/client-go/kubernetes"

	. "github.com/kedacore/keda/v2/tests/helper"
)

// Load environment variables from .env file
var _ = godotenv.Load("../../.env")

const (
	testName = "azure-log-analytics-test"
)

var (
	logAnalyticsWorkspaceID = os.Getenv("AZURE_LOG_ANALYTICS_WORKSPACE_ID")
	azureADClientID         = os.Getenv("AZURE_SP_APP_ID")
	azureADSecret           = os.Getenv("AZURE_SP_KEY")
	azureADTenantID         = os.Getenv("AZURE_SP_TENANT")
	testNamespace           = fmt.Sprintf("%s-ns", testName)
	secretName              = fmt.Sprintf("%s-secret", testName)
	deploymentName          = fmt.Sprintf("%s-deployment", testName)
	triggerAuthName         = fmt.Sprintf("%s-ta", testName)
	scaledObjectName        = fmt.Sprintf("%s-so", testName)
)

type templateData struct {
	TestNamespace           string
	SecretName              string
	DeploymentName          string
	TriggerAuthName         string
	ScaledObjectName        string
	AzureADClientID         string
	AzureADSecret           string
	AzureADTenantID         string
	LogAnalyticsWorkspaceID string
	QueryX, QueryY          int
}

type templateValues map[string]string

const (
	secretTemplate = `
apiVersion: v1
kind: Secret
metadata:
  name: {{.SecretName}}
  namespace: {{.TestNamespace}}
data:
  clientSecret: {{.AzureADSecret}}
`

	deploymentTemplate = `
apiVersion: apps/v1
kind: Deployment
metadata:
  name: {{.DeploymentName}}
  namespace: {{.TestNamespace}}
  labels:
    app: {{.DeploymentName}}
spec:
  replicas: 0
  selector:
    matchLabels:
      app: {{.DeploymentName}}
  template:
    metadata:
      labels:
        app: {{.DeploymentName}}
    spec:
      containers:
        - name: {{.DeploymentName}}
          image: nginx
`

	triggerAuthTemplate = `
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: {{.TriggerAuthName}}
  namespace: {{.TestNamespace}}
spec:
  secretTargetRef:
    - parameter: clientSecret
      name: {{.SecretName}}
      key: clientSecret
`

	scaledObjectTemplate = `
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: {{.ScaledObjectName}}
  namespace: {{.TestNamespace}}
  labels:
    deploymentName: {{.DeploymentName}}
spec:
  scaleTargetRef:
    name: {{.DeploymentName}}
  cooldownPeriod: 5
  pollingInterval: 5
  maxReplicaCount: 2
  triggers:
    - type: azure-log-analytics
      metadata:
        clientId: {{.AzureADClientID}}
        tenantId: {{.AzureADTenantID}}
        workspaceId: {{.LogAnalyticsWorkspaceID}}
        query: "let x = {{.QueryX}}; let y = {{.QueryY}}; print MetricValue = x, Threshold = y;"
        threshold: "1"
      authenticationRef:
        name: {{.TriggerAuthName}}
`
)

func TestScaler(t *testing.T) {
	// setup
	t.Log("--- setting up ---")
	require.NotEmpty(t, logAnalyticsWorkspaceID, "AZURE_LOG_ANALYTICS_WORKSPACE_ID env variable is required for deployment bus tests")
	require.NotEmpty(t, azureADClientID, "AZURE_SP_APP_ID env variable is required for deployment bus tests")
	require.NotEmpty(t, azureADSecret, "AZURE_SP_KEY env variable is required for deployment bus tests")
	require.NotEmpty(t, azureADTenantID, "AZURE_SP_TENANT env variable is required for deployment bus tests")

	// Create kubernetes resources
	kc := GetKubernetesClient(t)
	data, templates := getTemplateData()

	CreateKubernetesResources(t, kc, testNamespace, data, templates)

	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, 0, 60, 1),
		"replica count should be 0 after 1 minute")

	// test scaling
	testScaleUp(t, kc, data)
	testScaleDown(t, kc, data)

	// cleanup
	templates["triggerAuthTemplate"] = triggerAuthTemplate
	templates["scaledObjectTemplate"] = scaledObjectTemplate
	DeleteKubernetesResources(t, kc, testNamespace, data, templates)
}

func testScaleUp(t *testing.T, kc *kubernetes.Clientset, data templateData) {
	t.Log("--- testing scale up ---")
	data.QueryX = 10
	data.QueryY = 1

	KubectlApplyWithTemplate(t, data, "triggerAuthTemplate", triggerAuthTemplate)
	KubectlApplyWithTemplate(t, data, "scaledObjectTemplate", scaledObjectTemplate)

	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, 2, 60, 1),
		"replica count should be 2 after 1 minute")
}

func testScaleDown(t *testing.T, kc *kubernetes.Clientset, data templateData) {
	t.Log("--- testing scale down ---")
	data.QueryX = 0

	KubectlApplyWithTemplate(t, data, "scaledObjectTemplate", scaledObjectTemplate)

	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, 0, 60, 1),
		"replica count should be 0 after 1 minute")
}

func getTemplateData() (templateData, templateValues) {
	base64ClientSecret := base64.StdEncoding.EncodeToString([]byte(azureADSecret))

	return templateData{
			TestNamespace:           testNamespace,
			SecretName:              secretName,
			DeploymentName:          deploymentName,
			TriggerAuthName:         triggerAuthName,
			ScaledObjectName:        scaledObjectName,
			AzureADClientID:         azureADClientID,
			AzureADSecret:           base64ClientSecret,
			AzureADTenantID:         azureADTenantID,
			LogAnalyticsWorkspaceID: logAnalyticsWorkspaceID,
		}, templateValues{
			"secretTemplate":     secretTemplate,
			"deploymentTemplate": deploymentTemplate}
}
